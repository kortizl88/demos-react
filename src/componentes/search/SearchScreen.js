import React, { useMemo } from 'react'
import queryString from 'query-string'
import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../ui/heroes/HeroCard';
import getHeroesByName from '../../selectors/getHeroesByName';

export const SearchScreen = ({history}) => {

    const location = useLocation();

    const { q }  =  queryString.parse( location.search );

    console.log(q);

    const [formValues, handleInputChange, reset] = useForm( {
        textoBusqueda: q
    })

    const { textoBusqueda } = formValues;

    const heroesFiltered = useMemo(() => getHeroesByName(q), [q])

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${textoBusqueda}`)
    }

    return (
        <div>
            <h1>Search Screen</h1>
            <hr/>

            <div className='row'>
                <div className='col-5'>
                    <div>Search form</div>
                    <hr/>
                    <form onSubmit={handleSearch}>
                        <input 
                        type='text'
                        placeholder='busca tu heroe'
                        className='form-control'
                        name='textoBusqueda'
                        autoComplete="off"
                        value={textoBusqueda}
                        onChange={handleInputChange} />

                    <button 
                        type="submit"
                        className='btn m-1 btn-block btn-outline-primary'>
                        Buscar
                    </button>

                    </form>
                </div>

                <div className="col-7">
                    <h4>Resultados</h4>
                    <hr/>

                    {
                        heroesFiltered.map( hero => (
                            <HeroCard 
                                key={hero.id}
                                {...hero} />
                        ))    
                    }

                </div>

            </div>
        </div>
    )
}
