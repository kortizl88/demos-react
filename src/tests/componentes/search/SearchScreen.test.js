import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router-dom"
import { SearchScreen } from "../../../componentes/search/SearchScreen"

describe('Pruebas en el componente <SearchScreen />', () => {

    test('debe mostrarse el componente con valores por defecto', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={["/buscar"]}>
                <Route path='/buscar' component={SearchScreen} />
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();

        expect( wrapper.find(".row").exists()).toBe(true);
        
    })

    test('debe de mostrar a batman y el input con el valor del query string', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={["/buscar?q=batman"]}>
                <Route path='/buscar' component={SearchScreen} />
            </MemoryRouter>
        )

        expect( wrapper.find("input").prop('value') ).toBe("batman");
        expect(wrapper).toMatchSnapshot();

    })

    test('debe de mostrar un error si no se encuentra el hero', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={["/buscar?q=batmanasdasd"]}>
                <Route path='/buscar' component={SearchScreen} />
            </MemoryRouter>
        )

        expect( wrapper.find('HeroCard').exists() ).toBe(false);
        
    })

    test('debe de llamar el push del history', () => {

        const history = {
            push: jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={["/buscar"]}>
                <Route path='/buscar' component={ ()=> <SearchScreen history={history} /> } />
            </MemoryRouter>
        )

        wrapper.find('input').simulate("change", {
            target: {
                "name": "textoBusqueda",
                "value": "batman"
            }
        })

        wrapper.find('form').prop("onSubmit")({
            preventDefault(){}
        });

        expect(history.push).toHaveBeenCalledWith("?q=batman")

    })
    
    
    
    
    
})
