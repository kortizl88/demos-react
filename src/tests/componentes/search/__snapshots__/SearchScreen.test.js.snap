// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <SearchScreen /> debe de mostrar a batman y el input con el valor del query string 1`] = `ReactWrapper {}`;

exports[`Pruebas en el componente <SearchScreen /> debe mostrarse el componente con valores por defecto 1`] = `ReactWrapper {}`;
